src/coreclr/jit/lsra.cpp

```diff
@@ -1087,73 +1048,41 @@ void LinearScan::setBlockSequence()
             assert(!"Switch with single successor");
         }
 
-        for (unsigned succIndex = 0; succIndex < numSuccs; succIndex++)
+        for (BasicBlock* const succ : block->Succs(compiler))
         {
-            BasicBlock* succ = block->GetSucc(succIndex, compiler);
-            if (checkForCriticalOutEdge && succ->GetUniquePred(compiler) == nullptr)
+            if (checkForCriticalOutEdge && (succ->GetUniquePred(compiler) == nullptr))
             {
                 blockInfo[block->bbNum].hasCriticalOutEdge = true;
                 hasCriticalEdges                           = true;
                 // We can stop checking now.
-                checkForCriticalOutEdge = false;
-            }
-
-            if (isTraversalLayoutOrder() || isBlockVisited(succ))
-            {
-                continue;
-            }
-
-            // We've now seen a predecessor, so add it to the work list and the "readySet".
-            // It will be inserted in the worklist according to the specified traversal order
-            // (i.e. pred-first or random, since layout order is handled above).
-            if (!BlockSetOps::IsMember(compiler, readySet, succ->bbNum))
-            {
-                JITDUMP("\tSucc block: " FMT_BB, succ->bbNum);
-                addToBlockSequenceWorkList(readySet, succ, predSet);
-                BlockSetOps::AddElemD(compiler, readySet, succ->bbNum);
+                break;
             }
         }
+    };
 
-        // For layout order, simply use bbNext
-        if (isTraversalLayoutOrder())
-        {
-            nextBlock = block->Next();
-            continue;
-        }
+    JITDUMP("Start LSRA Block Sequence: \n");
+    for (unsigned i = 0; i < dfsTree->GetPostOrderCount(); i++)
+    {
+        visitBlock(blockSequence[i]);
+    }
 
-        while (nextBlock == nullptr)
+    // If the DFS didn't visit any blocks, add them to the end of blockSequence
+    if (dfsTree->GetPostOrderCount() < compiler->fgBBcount)
+    {
+        // Unvisited blocks are more likely to be at the back of the list, so iterate backwards
```

Comment:
What are the examples where there will be left over unvisited blocks?

Reply:
Any blocks that don't have `FlowEdges` into them won't be visited by the RPO traversal. For every example I looked at, this happens when we have throw blocks without any explicit predecessors. For example:
```
BBnum BBid ref try hnd preds           weight   [IL range]   [jump]                            [EH region]        [flags]
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1    [000..00D)-> BB06(0.5),BB02(0.5)     ( cond )                     i LIR hascall idxlen
BB02 [0008]  1       BB01                  0.25 [00D..???)-> BB07(0.01),BB03(0.99)   ( cond )                     LIR internal idxlen
BB03 [0010]  2       BB02,BB05             3.96 [00D..01C)-> BB05(0.5),BB04(0.5)     ( cond )                     i LIR loophead idxlen bwd
BB04 [0006]  1       BB03                  0.99 [???..???)-> BB05(1)                 (always)                     i LIR idxlen
BB05 [0004]  2       BB03,BB04             3.96 [01C..026)-> BB03(0.5),BB06(0.5)     ( cond )                     i LIR hascall idxlen bwd
BB06 [0019]  3       BB01,BB05,BB09        1    [026..031)                           (return)                     i LIR gcsafe newobj
BB07 [0011]  2       BB02,BB09             0.04 [00D..01C)-> BB09(0.5),BB08(0.5)     ( cond )                     i LIR loophead idxlen bwd
BB08 [0013]  1       BB07                  0.01 [???..???)-> BB09(1)                 (always)                     i LIR idxlen
BB09 [0014]  2       BB07,BB08             0.04 [01C..026)-> BB07(0.5),BB06(0.5)     ( cond )                     i LIR hascall idxlen bwd
BB10 [0020]  0                             0    [???..???)                           (throw )                     i LIR rare keep internal
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

`BB10` would've been removed in previous phases if it weren't for the `BBF_DONT_REMOVE` flag set on it. `StackLevelSetter`, which runs right before LSRA, has some logic for removing throw blocks that are no longer needed, even if they're flagged with `BBF_DONT_REMOVE`. So if we still see these blocks by the time we get to LSRA, I guess they're needed?

Reply:
Since such blocks will be cold (I assume), I am wondering if we should add logic to LSRA to quickly assign registers in such blocks, similar to the way we do for MinOpts, to save some TP time. I am not sure if there will be resolution added to the hot blocks because of this decision and if there is a way to add resolution too in cold blocks.